torch/csrc/distributed/rpc/profiler/server_process_global_profiler.cpp

```diff
@@ -0,0 +1,96 @@
+#include <shared_mutex>
+
+#include <torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+namespace profiler {
+namespace processglobal {
+
+namespace {
+class StateStackEntry;
+
+#if defined(__MACH__)
+// Compiler error: 'shared_timed_mutex' is unavailable: introduced in
+// macOS 10.12
+using mutexType = std::mutex;
+using rLockType = std::unique_lock<std::mutex>;
+using wLockType = std::unique_lock<std::mutex>;
+#else
+using mutexType = std::shared_timed_mutex;
+using rLockType = std::shared_lock<std::shared_timed_mutex>;
+using wLockType = std::unique_lock<std::shared_timed_mutex>;
+#endif
+
+mutexType currentStateStackEntryMutex;
+std::shared_ptr<StateStackEntry> currentStateStackEntryPtr = nullptr;
+
+class StateStackEntry {
+ public:
+  static std::shared_ptr<State> current() {
+    rLockType rlock(currentStateStackEntryMutex);
+
+    if (!currentStateStackEntryPtr) {
+      return nullptr;
+    }
+    return currentStateStackEntryPtr->profilerProcessGlobalStatePtr_;
+  }
+
+  static void push(std::shared_ptr<State> profilerProcessGlobalStatePtr) {
+    wLockType wlock(currentStateStackEntryMutex);
+
+    auto previousStateStackEntryPtr = currentStateStackEntryPtr;
+    currentStateStackEntryPtr = std::make_shared<StateStackEntry>();
+    currentStateStackEntryPtr->previousStateStackEntryPtr_ =
+        previousStateStackEntryPtr;
+    currentStateStackEntryPtr->profilerProcessGlobalStatePtr_ =
+        std::move(profilerProcessGlobalStatePtr);
+  }
+
+  static std::shared_ptr<State> pop() {
+    wLockType wlock(currentStateStackEntryMutex);
+
+    auto poppedStateStackEntryPtr = currentStateStackEntryPtr;
+    TORCH_INTERNAL_ASSERT(
+        poppedStateStackEntryPtr &&
+        poppedStateStackEntryPtr->profilerProcessGlobalStatePtr_);
+    currentStateStackEntryPtr =
+        poppedStateStackEntryPtr->previousStateStackEntryPtr_;
+    return poppedStateStackEntryPtr->profilerProcessGlobalStatePtr_;
+  }
+
+ private:
+  std::shared_ptr<State> profilerProcessGlobalStatePtr_{nullptr};
+  std::shared_ptr<StateStackEntry> previousStateStackEntryPtr_{nullptr};
+};
+
+} // namespace
+
+using namespace torch::autograd::profiler;
+
+void enableServer(const ProfilerConfig& new_config) {
+  TORCH_CHECK(
+      new_config.state == ProfilerState::CPU, "RPC module only support CPU");
```

Comment:
CPU -> profiling CPU state?

As we will be adding CUDA support through TensorPipe in v1.6

Reply:
Let me remove this assertion. CPU+GPU training is supported with TensorPipe.